cmd/kubeadm/app/util/config/common.go

```diff
@@ -126,7 +126,7 @@ func NormalizeKubernetesVersion(cfg *kubeadmapi.ClusterConfiguration) error {
 	versionInfo := componentversion.Get()
 	if isKubeadmPrereleaseVersion(&versionInfo, k8sVersion, mcpVersion) {
 		klog.V(1).Infof("WARNING: tolerating control plane version %s, assuming that k8s version %s is not released yet",
-			cfg.KubernetesVersion, mcpVersion)
+			cfg.KubernetesVersion, k8sVersion.WithPreRelease(""))
```

Comment:
this part of kubeadm about populating the version constants from `versionInfo` and validating them is certainly one of the confusing ones.

i think we can reword the warning to:
```
klog.V(1).Infof("WARNING: tolerating control plane version %s as a pre-release version, cfg.KubernetesVersion)
```

it's more vague, but it will give us (developers) the indication.


the problem here is the error thrown later, so we had to return early and skip the check.

```
if k8sVersion.LessThan(mcpVersion) {
...
```

`k8sVersion` can be `1.25.0-alpha..` and `k8sVersion.WithPreRelease("")` will result in `1.25.0`, but the idea of the warning was to highlight the case when the hardcoded `mpcVersion` can later fail the mentioned `LessThan` check

> // During the k8s release process, a kubeadm version in the main branch could be 1.23.0-pre,
	// while the 1.22.0 version is not released yet. The MinimumControlPlaneVersion validation
	// in such a case will not pass, since the value of MinimumControlPlaneVersion would be
	// calculated as kubeadm version - 1 (1.22) and k8sVersion would still be at 1.21.x

so in current versions the problem will happen when someone starts testing `1.26.0-alpha` while `1.25.0` is not out yet.
`mpcVersion = 1.25.0` and `k8sVersion = 1.24.0`.

Reply:
> k8sVersion can be 1.25.0-alpha.. and k8sVersion.WithPreRelease("") will result in 1.25.0, but the idea of the warning was to highlight the case when the hardcoded mpcVersion can later fail the mentioned LessThan check

> so in current versions the problem will happen when someone starts testing 1.26.0-alpha while 1.25.0 is not out yet.

1.26.0-alpha is not ever going to fail the minimum version check though?

you'd have to be testing a pre-release version older than the minimum version, but that version will be like v1.22.0-pre or v1.21.0-pre if it's lower than v1.22.0, the version comparison in our semver package is prerelease aware.

this code comment doesn't make sense, the version should be 1.22.0-pre when 1.22.0 is not released yet, not 1.23.0-pre while waiting for v1.22.0 ...?

Reply:
agree that we should probably just reword and drop printing the second version, but I'm concerned that the logic sounds confused here regarding pre-release versions, something sounds off with that comment and probably also needs clarifying, or else kubeadm is really jumping a version ahead in pre-release and IMHO we should really not do that 🤔

Reply:
i think the comment is accurate, because `k8sVersion` can come from the "latest stable" GCS marker unless the user is explicit about it.

example:
- kubeadm version is `1.26.0-pre` in the binary
- minimum control plane version is kubeadm version - 1 = `1.25.0` (without "pre")
- `k8sVersion` comes from the "latest stable" marker which is at `1.24.x`

this problem period of time happens when the `release-1.25` branch has not released `1.25.0`, "latest stable" is still at `1.24.x`, yet master was just tagged as `1.26.0-pre` and supposedly that's the kubeadm version in the binary that we are testing.

i might be missing something here myself, but we should be probably just simplify the warning as proposed earlier.

Reply:
ACK, done.

Reply:
alright, LGTMed

i think there a couple of problems here:
- kubeadm using a version from the internet as it's "dynamic default" version. that's probably one of the first things i complained about when i started digging in the project
- the k8s release process / tagging being confusing - 1.24.0 (latest stable), 1.25.0 (not released yet), master tagged as 1.26.0-pre

Reply:
agree on both, for the first I feel like we should just use the compiled version like kubectl etc.?

For the k8s release process, I totally forgot that we have a moment with a master branch v1.N+2 while v1.N+1 is branched but not released yet, I think we should only release branch after we release but ... 🤷‍♂️

Reply:
yeah, it might be a good idea to switch kubeadm to use the version from the binary as the control plane version "static default".

it would require changing the defaulting and might break assumptions for some users, but actually i think it's a good idea because the whole "fetch latest PATCH version from the internet with this kubeadm binary" can actually not work in some cases around e.g. changed control plane flags in a PATCH release.